### PR TITLE
Apply doctrine/cs v4

### PR DESCRIPTION
**Pull request by carusogabriel**
Excluded rules will be done separately as most of them contains BC Break or change of logic :)

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -20,8 +20,7 @@
         "mongodb/mongodb": "^1.1",
```
Comment (carusogabriel) [inline composer.json]: Looks like adding `composer.lock` affected MongoDB. Is there something to do, or should we not include it here?

**Comment by Ocramius** [comment]
The failure on `nightly` seems unrelated

**Comment by Ocramius** [comment]
Yeah, we need to bring this to green before merging :-\

**Comment by carusogabriel** [comment]
@Ocramius Removing `compose.lock` did solve :+1:

**Comment by Ocramius** [comment]
@carusogabriel can you also instruct scrutinizer to use our CS? It seems to pick up something older.

**Comment by carusogabriel** [comment]
@Ocramius Like https://github.com/doctrine/doctrine2/commit/9f572d?

File: lib/Doctrine/Common/Cache/CacheProvider.php [hunk 2]
```diff
@@ -182,7 +180,6 @@ private function getNamespacedId(string $id) : string
     /**
      * Returns the namespace cache key.
      *
```
Comment (Majkl578) [inline lib/Doctrine/Common/Cache/CacheProvider.php]: superfluous line

File: lib/Doctrine/Common/Cache/CacheProvider.php [hunk 3]
```diff
@@ -192,11 +189,10 @@ private function getNamespaceCacheKey() : string
     /**
      * Returns the namespace version.
      *
```
Comment (Majkl578) [inline lib/Doctrine/Common/Cache/CacheProvider.php]: superfluous line

File: lib/Doctrine/Common/Cache/CouchbaseCache.php [hunk 4]
```diff
@@ -2,28 +2,25 @@
...
-     * @var Couchbase|null
-     */
+    /** @var Couchbase|null */
     private $couchbase;
 
     /**
      * Sets the Couchbase instance to use.
      *
-     * @param Couchbase $couchbase
      *
```
Comment (Majkl578) [inline lib/Doctrine/Common/Cache/CouchbaseCache.php]: superfluous line

File: lib/Doctrine/Common/Cache/ExtMongoDBCache.php [hunk 5]
```diff
@@ -19,23 +22,17 @@
...
-    /**
-     * @var bool
-     */
+    /** @var bool */
     private $expirationIndexCreated = false;
 
     /**
-     * Constructor.
+     *
      *
```
Comment (Majkl578) [inline lib/Doctrine/Common/Cache/ExtMongoDBCache.php]: two superfluous lines

File: lib/Doctrine/Common/Cache/ExtMongoDBCache.php [hunk 6]
```diff
@@ -182,9 +180,7 @@ protected function doGetStats()
     /**
      * Check if the document is expired.
      *
-     * @param BSONDocument $document
      *
```
Comment (Majkl578) [inline lib/Doctrine/Common/Cache/ExtMongoDBCache.php]: two superfluous lines

File: lib/Doctrine/Common/Cache/FileCache.php [hunk 7]
```diff
@@ -2,12 +2,34 @@
...
+use function sprintf;
+use function strlen;
+use function strrpos;
+use function substr;
+use function tempnam;
+use function unlink;
+
 /**
  * Base file cache driver.
  *
```
Comment (Majkl578) [inline lib/Doctrine/Common/Cache/FileCache.php]: superfluous line

File: lib/Doctrine/Common/Cache/FileCache.php [hunk 8]
```diff
@@ -25,28 +47,20 @@
...
-    /**
-     * @var bool
-     */
+    /** @var bool */
     private $isRunningOnWindows;
 
     /**
-     * Constructor.
+     *
      *
```
Comment (Majkl578) [inline lib/Doctrine/Common/Cache/FileCache.php]: two superfluous lines

File: lib/Doctrine/Common/Cache/FileCache.php [hunk 9]
```diff
@@ -259,11 +269,10 @@ private function getIterator() : \Iterator
     /**
      * @param string $name The filename
      *
```
Comment (Majkl578) [inline lib/Doctrine/Common/Cache/FileCache.php]: superfluous line

File: lib/Doctrine/Common/Cache/FilesystemCache.php [hunk 10]
```diff
@@ -2,15 +2,22 @@
...
+use function fgets;
+use function fopen;
+use function is_file;
+use function serialize;
+use function time;
+use function unserialize;
+
 /**
  * Filesystem cache driver.
  *
```
Comment (Majkl578) [inline lib/Doctrine/Common/Cache/FilesystemCache.php]: superfluous line

File: lib/Doctrine/Common/Cache/LegacyMongoDBCache.php [hunk 11]
```diff
@@ -6,27 +6,27 @@
...
-    /**
-     * @var bool
-     */
+    /** @var bool */
     private $expirationIndexCreated = false;
 
     /**
-     * Constructor.
+     *
      *
```
Comment (Majkl578) [inline lib/Doctrine/Common/Cache/LegacyMongoDBCache.php]: two superfluous lines

File: lib/Doctrine/Common/Cache/LegacyMongoDBCache.php [hunk 12]
```diff
@@ -155,7 +156,6 @@ protected function doGetStats()
...
      * @param array $document
      *
```
Comment (Majkl578) [inline lib/Doctrine/Common/Cache/LegacyMongoDBCache.php]: superfluous line

File: lib/Doctrine/Common/Cache/MemcachedCache.php [hunk 13]
```diff
@@ -2,30 +2,22 @@
...
-     * @var Memcached|null
-     */
+    /** @var Memcached|null */
     private $memcached;
 
     /**
      * Sets the memcache instance to use.
      *
-     * @param Memcached $memcached
      *
```
Comment (Majkl578) [inline lib/Doctrine/Common/Cache/MemcachedCache.php]: superfluous line

File: lib/Doctrine/Common/Cache/MongoDBCache.php [hunk 14]
```diff
@@ -4,19 +4,19 @@
 
 use MongoCollection;
 use MongoDB\Collection;
+use const E_USER_DEPRECATED;
+use function trigger_error;
 
 /**
  * MongoDB cache provider.
  *
```
Comment (Majkl578) [inline lib/Doctrine/Common/Cache/MongoDBCache.php]: superfluous line

File: lib/Doctrine/Common/Cache/MongoDBCache.php [hunk 15]
```diff
@@ -33,15 +33,13 @@ class MongoDBCache extends CacheProvider
...
-    /**
-     * @var CacheProvider
-     */
+    /** @var CacheProvider */
     private $provider;
 
     /**
-     * Constructor.
+     *
      *
```
Comment (Majkl578) [inline lib/Doctrine/Common/Cache/MongoDBCache.php]: two superfluous lines

File: lib/Doctrine/Common/Cache/PhpFileCache.php [hunk 16]
```diff
@@ -90,7 +97,6 @@ protected function doSave($id, $data, $lifeTime = 0)
     }
 
     /**
-     * @param string $id
      *
```
Comment (Majkl578) [inline lib/Doctrine/Common/Cache/PhpFileCache.php]: superfluous line

File: lib/Doctrine/Common/Cache/RedisCache.php [hunk 17]
```diff
@@ -3,25 +3,23 @@
...
-     * @var Redis|null
-     */
+    /** @var Redis|null */
     private $redis;
 
     /**
      * Sets the redis instance to use.
      *
-     * @param Redis $redis
      *
```
Comment (Majkl578) [inline lib/Doctrine/Common/Cache/RedisCache.php]: superfluous line

File: lib/Doctrine/Common/Cache/RiakCache.php [hunk 18]
```diff
@@ -3,30 +3,29 @@
...
 
-    /**
-     * @var \Riak\Bucket
-     */
+    /** @var Bucket */
     private $bucket;
 
     /**
      * Sets the riak bucket instance to use.
      *
```
Comment (Majkl578) [inline lib/Doctrine/Common/Cache/RiakCache.php]: superfluous line

File: lib/Doctrine/Common/Cache/RiakCache.php [hunk 19]
```diff
@@ -182,9 +181,7 @@ protected function doGetStats()
     /**
      * Check if a given Riak Object have expired.
      *
-     * @param \Riak\Object $object
      *
```
Comment (Majkl578) [inline lib/Doctrine/Common/Cache/RiakCache.php]: two superfluous lines

File: lib/Doctrine/Common/Cache/SQLite3Cache.php [hunk 20]
```diff
@@ -4,48 +4,50 @@
...
+use function array_search;
+use function implode;
+use function serialize;
+use function sprintf;
+use function time;
+use function unserialize;
 
 /**
  * SQLite3 cache provider.
  *
```
Comment (Majkl578) [inline lib/Doctrine/Common/Cache/SQLite3Cache.php]: superfluous line

File: lib/Doctrine/Common/Cache/SQLite3Cache.php [hunk 21]
```diff
@@ -4,48 +4,50 @@
...
-    /**
-     * @var string
-     */
+    /** @var string */
     private $table;
 
     /**
-     * Constructor.
+     *
      *
```
Comment (Majkl578) [inline lib/Doctrine/Common/Cache/SQLite3Cache.php]: two superfluous lines

File: lib/Doctrine/Common/Cache/SQLite3Cache.php [hunk 22]
```diff
@@ -199,7 +200,6 @@ private function getFields() : array
...
      * @param array $item
      *
```
Comment (Majkl578) [inline lib/Doctrine/Common/Cache/SQLite3Cache.php]: superfluous line

File: phpcs.xml.dist [hunk 23]
```diff
@@ -14,9 +14,22 @@
...
     <file>tests</file>
 
-    <rule ref="Doctrine"/>
+    <rule ref="Doctrine">
+        <exclude name="SlevomatCodingStandard.TypeHints.DeclareStrictTypes"/>
+        <exclude name="SlevomatCodingStandard.TypeHints.TypeHintDeclaration.MissingParameterTypeHint"/>
+        <exclude name="SlevomatCodingStandard.TypeHints.TypeHintDeclaration.MissingReturnTypeHint"/>
+        <exclude name="SlevomatCodingStandard.ControlStructures.DisallowEqualOperators"/>
+        <exclude name="SlevomatCodingStandard.TypeHints.TypeHintDeclaration"/>
+        <exclude name="Generic.CodeAnalysis.EmptyStatement"/>
```
Comment (Majkl578) [inline phpcs.xml.dist]: Why?

File: phpcs.xml.dist [hunk 24]
```diff
@@ -14,9 +14,22 @@
...
+        <exclude name="SlevomatCodingStandard.ControlStructures.EarlyExit"/>
```
Comment (Majkl578) [inline phpcs.xml.dist]: Why?

**Comment by carusogabriel** [comment]
@Majkl578 Applied you requested changes at 5f92b0b. Please review before I squash to merge. The left excluded rules will be done separately, as some of them are BC like adding return types, param types, `strict_types`, or have too many cases and needs a careful review :)

**Comment by Majkl578** [comment]
I understand why return/parameter types etc. are excluded, I only asked for two specific: EarlyExit and EmptyStatement. :innocent:

**Comment by Ocramius** [comment]
LGTM :+1:

**Comment by nicolas-grekas** [comment]
This PR must embed a regression, Symfony tests are red because of it.

**Comment by nicolas-grekas** [comment]
Eg on branch 3.4:
```

./phpunit src/Symfony/Component/Cache/ --filter Doctrine
Testing src/Symfony/Component/Cache/
FFSS.............S...S.........................................  63 / 326 ( 19%)
............................................................... 126 / 326 ( 38%)
.FSS.F......F.................................................. 189 / 326 ( 57%)
.................S............................................. 252 / 326 ( 77%)
............................................................... 315 / 326 ( 96%)
..........S                                                     326 / 326 (100%)

Time: 385 ms, Memory: 14.00MB

There were 5 failures:

1) Symfony\Component\Cache\Tests\Adapter\DoctrineAdapterTest::testDefaultLifeTime
Failed asserting that true is false.

/home/nicolas/Code/symfony-symfony/src/Symfony/Component/Cache/Tests/Adapter/AdapterTestCase.php:51

2) Symfony\Component\Cache\Tests\Adapter\DoctrineAdapterTest::testExpiration
Failed asserting that true is false.

/home/nicolas/Code/symfony-symfony/src/Symfony/Component/Cache/Tests/Adapter/AdapterTestCase.php:66

3) Symfony\Component\Cache\Tests\Simple\DoctrineCacheTest::testDefaultLifeTime
Failed asserting that 'value' is null.

/home/nicolas/Code/symfony-symfony/src/Symfony/Component/Cache/Tests/Simple/CacheTestCase.php:52

4) Symfony\Component\Cache\Tests\Simple\DoctrineCacheTest::testSetTtl
Value must expire after ttl.
Failed asserting that 'value' is null.

/home/nicolas/Code/symfony-symfony/vendor/cache/integration-tests/src/SimpleCacheTest.php:146

5) Symfony\Component\Cache\Tests\Simple\DoctrineCacheTest::testSetMultipleTtl
Value must expire after ttl.
Failed asserting that 'value2' is null.

/home/nicolas/Code/symfony-symfony/vendor/cache/integration-tests/src/SimpleCacheTest.php:240
```

**Comment by Ocramius** [comment]
@Majkl578 I'd love to bisect this one <_<

**Comment by nicolas-grekas** [comment]
Hum, see #253

**Pull request merged**